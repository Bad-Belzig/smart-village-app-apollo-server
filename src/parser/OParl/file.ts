import { isObjectLike } from 'lodash';

import { parseDate } from '../dateParser';

// only the necessary validation for parsing is done here.
// the created database object will be validated after creation before saving
// by using the .validate method that checks against the provided schema
export const parseFile = (json) => {
  if (isObjectLike(json)) {
    const externalId = json.id; // change the id given to us to be the externalId instead
    delete json.id; // remove the id key from the json to ensure we do not preset it for mongo db
    return {
      ...json,
      externalId,
      created: parseDate(json.created),
      modified: parseDate(json.modified),

      date: parseDate(json.date),
    };
  }
};
