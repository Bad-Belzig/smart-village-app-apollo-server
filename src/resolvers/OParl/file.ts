import { FilterQuery } from 'mongoose';

import {
  AgendaItem,
  File,
  IFile,
  IFileSchema,
  Meeting,
  Paper,
} from '../../models';
import { getPaginatedEntriesByIds } from '../resolverHelpers';

export const fileResolvers = {
  Query: {
    oParlFiles: (_, args: { externalIds?: string[]; keyword?: string[] }) => {
      const filter: FilterQuery<IFileSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };

      return File.find(filter);
    },
  },
  OParlFile: {
    masterFile: (parent: IFile) =>
      File.findOne({ externalId: parent.masterFile }),
    derivativeFile: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        File,
        parent.derivativeFile,
        args.offset,
        args.pageSize,
      ),
    meeting: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Meeting,
        parent.meeting,
        args.offset,
        args.pageSize,
      ),
    agendaItem: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        AgendaItem,
        parent.agendaItem,
        args.offset,
        args.pageSize,
      ),
    paper: async (
      parent: IFile,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(Paper, parent.paper, args.offset, args.pageSize),
  },
};
