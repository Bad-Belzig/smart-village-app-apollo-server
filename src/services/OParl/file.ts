import { fetch } from 'apollo-env';

import { File } from '../../models';
import { parseFile } from '../../parser';

// import flow
// 1. fetch json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importFile = async (url: string) => {
  try {
    const response = await fetch(url);

    if (response.ok) {
      const json = await response.json();

      // TODO: check for previous existence

      const lt = new File(parseFile(json));

      await lt.validate();

      // TODO: check for files, meetings, agenda items and papers to import

      return lt.save();
    } else {
      throw new Error(
        `Error while fetching File from : ${url}! \n response.status: ${response.status}`,
      );
    }
  } catch (e) {
    console.log(e);
  }
};
