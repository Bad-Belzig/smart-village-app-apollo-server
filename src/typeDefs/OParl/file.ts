import { gql } from 'apollo-server-core';

export const fileGQL = gql`
  extend type Query {
    oParlFiles: [OParlFile!]
  }

  type OParlFile {
    id: ID!
    "id from the oparl object = url where the json is located"
    externalId: String!
    "oParl type = url where the type is specified"
    type: String!
    created: Date
    modified: Date
    keyword: [String!]
    web: String
    deleted: Boolean
    accessUrl: String!
    agendaItem: [OParlAgendaItem!]
    date: Date
    derivativeFile: [OParlFile!]
    downloadUrl: String
    externalServiceUrl: String
    fileLicense: String
    fileName: String
    masterFile: OParlFile
    meeting: [OParlMeeting!]
    mimeType: String
    name: String
    paper: [OParlPaper!]
    sha1Checksum: String
    size: Int
    text: String
  }
`;
