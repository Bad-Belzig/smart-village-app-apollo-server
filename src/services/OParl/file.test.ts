import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../testHelpers';
import { importFile } from './file';

const helperFuncs = setupTestDataBase('test_file_import');

// TODO: change test url to something within our control, that uses all fields
const testUrl =
  'https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/files.asp?dtyp=110&id=21740';

describe('importing a file', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importFile, testUrl));

  it('import returns undefined on faulty url', invalidUrlTest(importFile));

  afterAll(helperFuncs.afterAll);
});
