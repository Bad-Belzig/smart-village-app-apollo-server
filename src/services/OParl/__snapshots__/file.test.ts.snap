// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a file imports properly 1`] = `
Object {
  "__v": 0,
  "accessUrl": "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/download.asp?dtyp=110&id=21740",
  "agendaItem": Array [],
  "created": 2011-11-02T11:00:00.000Z,
  "derivativeFile": Array [],
  "externalId": "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/files.asp?dtyp=110&id=21740",
  "fileName": "21740.pdf",
  "keyword": Array [],
  "meeting": Array [],
  "mimeType": "application/pdf",
  "modified": 2017-02-09T15:15:31.000Z,
  "name": "Niederschrift",
  "paper": Array [],
  "size": 334411,
  "type": "https://schema.oparl.org/1.0/File",
}
`;
